ps2-decompile/attach.c

```diff
@@ -3,188 +3,170 @@
-                    } else {
-                        printf("sceUsbdScanStaticDescriptor (HID(0x21)) failed\n");
-                        result = -1;
-                    }
-                } else {
-                    printf("sceUsbdScanStaticDescriptor (ENDPOINT) failed\n");
-                    result = -1;
-                }
-            } else {
-                printf("wrong bNumEndpoints : %d\n", idesc->bNumEndpoints);
-                result = -1;
-            }
-        } else {
-            printf("sceUsbdScanStaticDescriptor (INTERFACE) failed\n");
-            result = -1;
-        }
-    } else {
+    if ((cdesc = sceUsbdScanStaticDescriptor(devId, 0, USB_DT_CONFIG)) == NULL) {
         printf("sceUsbdScanStaticDescriptor (CONFIGURATION) failed\n");
-        result = -1;
+        return -1;
+    }
+
+    if ((idec = sceUsbdScanStaticDescriptor(devId, cdesc, USB_DT_INTERFACE)) == NULL) {
+        printf("sceUsbdScanStaticDescriptor (INTERFACE) failed\n");
+        return -1;
+    }
+
+    if (idesc->bNumEndpoints != 1) {
+        printf("wrong bNumEndpoints : %d\n", idesc->bNumEndpoints);
+        return -1;
     }
-    return result;
+
+    if ((edesc = sceUsbdScanStaticDescriptor(devId, idesc, USB_DT_ENDPOINT)) == NULL) {
+        printf("sceUsbdScanStaticDescriptor (ENDPOINT) failed\n");
+        return -1;
+    }
+
+    if ((hdesc = sceUsbdScanStaticDescriptor(devId, 0, 33)) == NULL) {
+        printf("sceUsbdScanStaticDescriptor (HID(0x21)) failed\n");
+        return -1;
+    }
+
+    if ((edesc->bEndpointAddress & 0x80) == 0) {
+        return -1;
+    }
+
+    if ((edesc->bmAttributes & 3) != 3) {
+        return -1;
+    }
+
+    if ((maxPacketSize = edesc->wMaxPacketSizeLB | (edesc->wMaxPacketSizeHB << 8)) >= 9) {
+        printf("wrong payload size %d\n", maxPacketSize);
+        return -1;
+    }
+
+    if ((g = unit_alloc(devId, maxPacketSize,idesc->bInterfaceNumber, idesc->bAlternateSetting)) == NULL) {
+        printf("unit_alloc failed\n");
+        return -1;
+    }
+
+    if ((g->configEnpoint = sceUsbdOpenPipe(devId, 0)) < 0) {
+        printf("sceUsbdOpenPipe failed(1)\n");
+        unit_free(g);
+        return -1;
+    }
+
+    printf("#### sceUsbdOpenPipe %d\n", g->configEndpoint);
+
+    if ((g->dataEndpoint = sceUsbdOpenPipe(devId, edesc)) < 0) {
+        printf("sceUsbdOpenPipe failed(2)\n");
+        unit_free(g);
+        return -1;
+    }
+
+    sceUsbdSetPrivateData(devId, g);
+    g->descriptorChecksPassed = 1;
+
+    if (cdesc->bLength != 9 ||
+        LOBYTE(cdesc->wTotalLength) != 0x22 || HIBYTE(cdesc->wTotalLength) ||
+        cdesc->bNumInterfaces != 1 || cdesc->bConfigurationValue != 1 ||
+        cdesc->iConfiguration || cdesc->bmAttributes != 0x80 ||
+        cdesc->maxPower != 0xC8) {
+        g->descriptorChecksPassed = 0;
+        printf("cdesc check failed\n");
+    }
+    if (hdesc->bLength != 9 ||
+        hdesc->bDescriptorType != 0x21 ||
+        LOBYTE(hdesc->bcdUSB) != 0x10 ||
```

Comment:
This can be converted to compare both bytes at once. `hdec->bcdUSB == 0x110` from `((1 << 8) | 0x10)`. `LOBYTE` and `HIBYTE` macros (and similar such as a `LOWORD`) should be removed and instead compare the whole value.

`(hb << 8) | lb`